"Task Completed" page & rating functionality

nicholaspad opened the pull request:
## Summary

### Task Completed page

- Redesigned Task Completed page to include more prominent crypto reward and a more compact rating input
- Taskers can now click a star count to submit a task rating (currently irreversible; they are asked to confirm their selection before it posts)
    - After a rating is submitted, the stars become read-only. Reloading the page will result in a denial of access and auto-redirect to Browse Tasks.
- Submitting a task now routes to `/tasker/task-completed/<taskId>`
- Accessing a Task Completed page requires the task to have a status of "pending verification" and requires the Tasker to have not yet rated the task

### Average rating computation

- Average task rating is now dynamically computed and displayed in the Browse Tasks table
- Average Requester rating is now dynamically computed and displayed in the Task Overview/Details pages
    - This is the average of all the Requester's created tasks' ratings
- Removed `avgRating` field from `Tasks` collection and `requesterRating` field from the `Users` collection

### Other

- Moved helper functions from `cloud/cloud.js` to `cloud/helpers.js`

**See additional changes comment below**

## Demos

### Full experience demonstration

![Apr-24-2022 00-10-44](https://user-images.githubusercontent.com/13815069/164956259-5fb2c8fd-9b60-490c-8e79-a23591534fbd.gif)

jennsun commented on cloud/helpers.js:
```diff
@@ -0,0 +1,126 @@
+      value: maxReward,
+    })
+  );
+  const unitRewardWei = maxRewardWei / maxTaskers - 1; // ensures that unitRewardWei*maxTaskers < maxRewardWei
```
could this potentially be the source of the reward discrepancy between the MoralisDB and smart contract? Seems like unitRewardWei is rounded down

christine-sun commented on cloud/helpers.js:
```diff
@@ -0,0 +1,126 @@
+      value: maxReward,
+    })
+  );
+  const unitRewardWei = maxRewardWei / maxTaskers - 1; // ensures that unitRewardWei*maxTaskers < maxRewardWei
```
Just a note that the discrepancy between this and the smart contract is here, where the unitRewardWei was calculated without the -1 in the smart contract. @jennsun let's modify this to make the two consistent

nicholaspad replied:
Shouldn't this cause a discrepancy of 1 Wei = 10^-18 MATIC?

jennsun commented on pages/tasker/task-completed/[taskId].tsx:
```diff
@@ -0,0 +1,200 @@
+        `Are you sure you want to rate this task as ${rating} stars? You may rate only once!`
+      )
+    ) {
+      setRating(0);
```
can you confirm the rating should be set to 0 here and not rating?

jennsun commented on pages/browse-tasks.tsx:
```diff
@@ -60,14 +60,16 @@ const extraColumns: GridColDef[] = [
-          icon={<StarRoundedIcon fontSize="inherit" />}
-          emptyIcon={<StarOutlineRoundedIcon fontSize="inherit" />}
-        />
+        <Box ml="-7px">
```
hm do we have to have a negative margin? honestly a nit so I don't think we need to worry about it

nicholaspad replied:
Yes, without the negative margin the stars look too far to the right!

jennsun reviewed (changes requested):
had 2 questions, otherwise this looks good!

christine-sun commented on cloud/helpers.js:
```diff
@@ -0,0 +1,126 @@
+    countRating++;
+  });
+  if (countRating === 0) return -1;
+  return avgRating / countRating;
```
Do we want to store the raw number or round it?

nicholaspad replied:
This number actually isn't being stored in the database, it's computed on request whenever a page needs it

christine-sun commented on pages/browse-tasks.tsx:
```diff
@@ -60,14 +60,16 @@ const extraColumns: GridColDef[] = [
+            readOnly
+            value={params.row.rating}
+            size="large"
+            precision={0.5}
```
I made an earlier comment about rounding the rating and see here that there's this precision variable. Wanted to clarify if this is doing the "rounding" visually here closer to rendering?

nicholaspad replied:
Yes, this is doing the rounding to the nearest 0.5

christine-sun commented on pages/tasker/task-completed/[taskId].tsx:
```diff
@@ -0,0 +1,200 @@
+                </>
+              ) : (
+                <>
+                  <Typography
+                    variant="h6"
+                    fontWeight="normal"
+                    color="primary"
+                    mr={2}
+                  >
+                    Rate your experience:
```
Could we give this text the secondary purple color?

christine-sun commented on pages/tasker/task-completed/[taskId].tsx:
```diff
@@ -0,0 +1,200 @@
+              }}
+            >
+              <Typography color="primary" fontWeight={400} fontSize={20}>
+                {"5"} ETH
```
Note to pull actual reward amount

nicholaspad replied:
will do, i forgot about this hardcode lol

christine-sun reviewed (approved):
Looks really great overall!

nicholaspad commented on pages/tasker/task-completed/[taskId].tsx:
```diff
@@ -0,0 +1,200 @@
+              }}
+            >
+              <Typography color="primary" fontWeight={400} fontSize={20}>
+                {"5"} ETH
```
Oops, will fix that lol

nicholaspad commented:
## Changes

@jennsun @christine-sun 

- Added optional truncation value for table cells - if text is truncated, an ellipsis is appended and a tooltip is added automatically (see demo below)
- Fixed hardcoded crypto reward value on Task Completed pages
- Centered crypto reward value text on Task Completed pages

![Apr-26-2022 14-11-29](https://user-images.githubusercontent.com/13815069/165365418-0028b269-23c7-4315-b087-7bc694755fa0.gif)

jennsun commented:
The tooltip is a great idea! Thanks for the changes Nick, they look great
